intermediate_source/TP_tutorial.rst

```diff
@@ -0,0 +1,282 @@
+
+In a typical TransformerBlock, the forward function combines norm layers (``LayerNorm`` or ``RMSNorm``), an attention layer, a feedforward layer, and residual connections. For example:
+
+.. code-block:: python
+    # forward in a TransformerBlock
+    def forward(self, x):
+        h = x + self.attention(self.attention_norm(x))
+        out = h + self.feed_forward(self.ffn_norm(h))
+        return out
+
+In most use cases, the activations (and gradients) are of the shape ``[batch size, sequence length, hidden dimension]`` outside the Attention and FeedForward modules. In the DTensor’s language, Sequence Parallelism maintains a Shard(1) layout for the activations, and does conversions as needed when entering or exiting those modules. Following the code example earlier, the code below demonstrates how we apply Sequence Parallel to the norm layers within a TransformerBlock:
+
+.. code-block:: python
+    from torch.distributed.tensor.parallel import (
+        ColwiseParallel,
+        PrepareModuleInput,
+        RowwiseParallel,
+        SequenceParallel,
+        parallelize_module
+    )
+
+    tp_plan = {
+        # by default the input and output of SequenceParallel
+        # has Shard(1) layouts
+        "attention": PrepareModuleInput(
+            input_layouts=(Shard(1),),
+            desired_input_layouts=(Replicate(),),
+        ),
+        "attention.wq": ColwiseParallel(),
+        "attention.wk": ColwiseParallel(),
+        "attention.wv": ColwiseParallel(),
+        "attention.wo": RowwiseParallel(output_layouts=Shard(1)),
+        ​​"attention_norm": SequenceParallel(),
+        "feed_forward": PrepareModuleInput(
+            input_layouts=(Shard(1),),
+            desired_input_layouts=(Replicate(),),
+        ),
+        "feed_foward.w1": ColwiseParallel(),
+        "feed_forward.w2": RowwiseParallel(output_layouts=Shard(1)),
+        "feed_forward.w3": ColwiseParallel(),
+        "ffn_norm": SequenceParallel(),
+    }
+
+One can see we now use PrepareModuleInput to modify the module input layouts to the Attention and FeedForward layers from ``Shard(1)`` to ``Replicate()``, and mark their output layouts as ``Shard(1)``. Just like what happens to Tensor Parallelism, one only needs to specify the tensor sharding layouts of the inputs and outputs, and the communication between layers will happen automatically.
+
+Note that with Sequence Parallel, we assume the inputs and outputs of a TransformerBlock are always sharded on the sequence dimension, so that multiple TransformerBlocks can be concatenated seamlessly. The only exception is that the input to the first TransformerBlock is replicated, so it has to be converted explicitly:
```

Comment:
```suggestion
Note that with Sequence Parallel, we assume the inputs and outputs of a ``TransformerBlock`` are always sharded on the sequence dimension, so that multiple ``TransformerBlocks`` can be concatenated seamlessly. The only exception is that the input to the first ``TransformerBlock`` is replicated, so it has to be converted explicitly:
```